llvm/lib/CheerpWriter/JSInterop.cpp

```diff
@@ -38,6 +39,86 @@ uint32_t CheerpWriter::countJsParameters(const llvm::Function* F, bool isStatic)
 	return ret;
 }
 
+static std::pair<std::string, std::string> buildArgumentsString(const llvm::Function* F, bool isStatic, const PointerAnalyzer& PA, const std::map<const StructType*, StringRef>& toBeWrapped)
+{
+	// While code-generating JSExported functions we need to write something like:
+	// function someName(a0,a1,a2,a3,a4,a5) {
+	// 	return equivalentInnerFunctionName(a0,a1,a2,[a3],0,a4,a5);	//very much an example, where a3 needs to be passes as SPLIT_REGULAR
+	// }
+	//
+	// This helper function will return (via a pair) two strings that will be used like:
+	// function someName(PAIR_FIRST_MEMBER)
+	// {
+	// 	return equivalentInnerFunctionName(PAIR_SECOND_MEMBER);
+	// }
+
+
+	std::ostringstream args_outer;		// args_outer will contain the comma-separated list of arguments to the outer function
+	std::ostringstream args_inner;		// args_inner will contain the comma-separated list of arguments to the inner function
+
+	uint32_t ret = 0;
+	auto it = F->arg_begin();
+	auto itE = F->arg_end();
+	if(!isStatic)
+	{
+		// The this parameter is implicit, skip it
+		++it;
+	}
+	while(it != itE)
+	{
+		std::ostringstream param;
+		param << "a" << ret;
+		const std::string aX = param.str();
+
+		bool isWrapped = false;
+		if (it->getType()->isPointerTy() && toBeWrapped.count(dyn_cast<StructType>(it->getType()->getPointerElementType())))
+			isWrapped = true;
+
+		args_inner << aX << ",";
+		if(it->getType()->isPointerTy() && PA.getPointerKind(&(*it)) == SPLIT_REGULAR)
+		{
+			if (isWrapped)
+			{
+				// Split regular representation
+				args_outer << aX << ".this.d," << aX << ".this.o,";
+			}
+			else
+			{
+				// Implicitly split regular
+				args_outer << "[" << aX << "]," << "0,";
```

Comment:
This is not correct. If we create a SPLIT_REGULAR like this it will lose the property of comparing equal to other SPLIT_REGULARs of the same object. Moreover, how can this ever happen?

Reply:
This is wrong, and somehow was untested, have a fix but first would like to find a testcase that check this.